File: Java/Problem_056_MergeIntervals.java
/*
Problem: 56 - Merge Intervals
Difficulty: Medium
Link: https://leetcode.com/problems/merge-intervals/

Problem Statement:
Given an array of intervals where intervals[i] = [start_i, end_i], 
merge all overlapping intervals, and return an array of the non-overlapping 
intervals that cover all the intervals in the input.

Approach:
1. Sort intervals by start time
2. Iterate through sorted intervals
3. If current interval overlaps with last merged interval, merge them
   (overlap exists if: current_start <= last_end)
4. Otherwise, add current interval to result

Two intervals [a,b] and [c,d] overlap if: c <= b

Time Complexity: O(N log N) - Dominated by sorting
Space Complexity: O(N) - For storing result (or O(log N) for sorting)
*/

import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {
        // Edge case: empty or single interval
        if (intervals.length <= 1) {
            return intervals;
        }
        
        // Sort intervals by start time
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        // List to store merged intervals
        List<int[]> merged = new ArrayList<>();
        
        // Add first interval
        int[] currentInterval = intervals[0];
        merged.add(currentInterval);
        
        // Process each interval
        for (int i = 1; i < intervals.length; i++) {
            int currentStart = intervals[i][0];
            int currentEnd = intervals[i][1];
            int lastEnd = currentInterval[1];
            
            // Check if intervals overlap
            if (currentStart <= lastEnd) {
                // Merge: extend the end of current interval
                currentInterval[1] = Math.max(lastEnd, currentEnd);
            } else {
                // No overlap: add new interval
                currentInterval = intervals[i];
                merged.add(currentInterval);
            }
        }
        
        // Convert list to array
        return merged.toArray(new int[merged.size()][]);
    }
    
    // Alternative: Using LinkedList for better performance
    public int[][] mergeLinkedList(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }
        
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        
        LinkedList<int[]> merged = new LinkedList<>();
        
        for (int[] interval : intervals) {
            // If list is empty or no overlap
            if (merged.isEmpty() || merged.getLast()[1] < interval[0]) {
                merged.add(interval);
            } else {
                // Merge with last interval
                merged.getLast()[1] = Math.max(merged.getLast()[1], interval[1]);
            }
        }
        
        return merged.toArray(new int[merged.size()][]);
    }
    
    // Helper method to print 2D array
    public static void printIntervals(int[][] intervals) {
        System.out.print("[");
        for (int i = 0; i < intervals.length; i++) {
            System.out.print("[" + intervals[i][0] + "," + intervals[i][1] + "]");
            if (i < intervals.length - 1) {
                System.out.print(",");
            }
        }
        System.out.println("]");
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        int[][] intervals1 = {{1,3}, {2,6}, {8,10}, {15,18}};
        System.out.print("Input: ");
        printIntervals(intervals1);
        int[][] result1 = solution.merge(intervals1);
        System.out.print("Output: ");
        printIntervals(result1);
        System.out.println("Expected: [[1,6],[8,10],[15,18]]\n");
        
        // Test case 2
        int[][] intervals2 = {{1,4}, {4,5}};
        System.out.print("Input: ");
        printIntervals(intervals2);
        int[][] result2 = solution.merge(intervals2);
        System.out.print("Output: ");
        printIntervals(result2);
        System.out.println("Expected: [[1,5]]\n");
        
        // Test case 3
        int[][] intervals3 = {{1,4}, {0,4}};
        System.out.print("Input: ");
        printIntervals(intervals3);
        int[][] result3 = solution.merge(intervals3);
        System.out.print("Output: ");
        printIntervals(result3);
        System.out.println("Expected: [[0,4]]\n");
        
        // Test case 4
        int[][] intervals4 = {{1,4}, {2,3}};
        System.out.print("Input: ");
        printIntervals(intervals4);
        int[][] result4 = solution.merge(intervals4);
        System.out.print("Output: ");
        printIntervals(result4);
        System.out.println("Expected: [[1,4]]\n");
        
        // Test case 5: Multiple overlapping
        int[][] intervals5 = {{1,3}, {2,4}, {3,5}, {6,7}};
        System.out.print("Input: ");
        printIntervals(intervals5);
        int[][] result5 = solution.mergeLinkedList(intervals5);
        System.out.print("Output (LinkedList): ");
        printIntervals(result5);
        System.out.println("Expected: [[1,5],[6,7]]");
    }
}

File: Python/Problem_056_merge_intervals.py
"""
LeetCode Problem: 56 - Merge Intervals
Difficulty: Medium
Link: https://leetcode.com/problems/merge-intervals/

Problem Statement:
Given an array of intervals where intervals[i] = [start_i, end_i], merge all 
overlapping intervals, and return an array of the non-overlapping intervals 
that cover all the intervals in the input.

Examples:
Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].

Input: intervals = [[1,4],[4,5]]
Output: [[1,5]]
Explanation: Intervals [1,4] and [4,5] are considered overlapping.

Constraints:
- 1 <= intervals.length <= 10^4
- intervals[i].length == 2
- 0 <= start_i <= end_i <= 10^4
"""

from typing import List


class Solution:
    """
    Approach: Sort and Merge
    
    Time Complexity: O(n log n) - dominated by sorting
    Space Complexity: O(n) - for output array (or O(log n) if considering sort space)
    
    Key Insights:
    - Sort intervals by start time
    - After sorting, overlapping intervals are adjacent
    - Two intervals [a,b] and [c,d] overlap if c <= b (when a <= c)
    - Merge by taking min(start) and max(end) of overlapping intervals
    """
    
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        """
        Merge all overlapping intervals.
        
        Args:
            intervals: List of [start, end] pairs
            
        Returns:
            List of merged non-overlapping intervals
        """
        # Edge case: empty or single interval
        if len(intervals) <= 1:
            return intervals
        
        # Sort by start time
        intervals.sort(key=lambda x: x[0])
        
        merged = []
        current_start, current_end = intervals[0]
        
        for start, end in intervals[1:]:
            # Check if current interval overlaps with next
            if start <= current_end:
                # Merge: extend current_end to max of both ends
                current_end = max(current_end, end)
            else:
                # No overlap: save current and start new interval
                merged.append([current_start, current_end])
                current_start, current_end = start, end
        
        # Don't forget the last interval
        merged.append([current_start, current_end])
        
        return merged


class SolutionAlternative:
    """
    Alternative: Using stack-like approach with result array
    
    Same complexity but slightly different style
    """
    
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        """Merge using result array as stack."""
        if not intervals:
            return []
        
        intervals.sort()
        merged = [intervals[0]]
        
        for start, end in intervals[1:]:
            # Check if overlaps with last interval in merged
            if start <= merged[-1][1]:
                # Merge by updating end of last interval
                merged[-1][1] = max(merged[-1][1], end)
            else:
                # No overlap: add new interval
                merged.append([start, end])
        
        return merged


# TEST CASES

def test_solution():
    """Test cases with assertions."""
    sol = Solution()
    
    # Test case 1: Example from problem
    assert sol.merge([[1,3],[2,6],[8,10],[15,18]]) == [[1,6],[8,10],[15,18]], "Test 1 failed"
    
    # Test case 2: Adjacent intervals (touching)
    assert sol.merge([[1,4],[4,5]]) == [[1,5]], "Test 2 failed"
    
    # Test case 3: Single interval
    assert sol.merge([[1,4]]) == [[1,4]], "Test 3 failed"
    
    # Test case 4: No overlaps
    assert sol.merge([[1,2],[3,4],[5,6]]) == [[1,2],[3,4],[5,6]], "Test 4 failed"
    
    # Test case 5: All overlap into one
    assert sol.merge([[1,4],[2,3]]) == [[1,4]], "Test 5 failed"
    
    # Test case 6: Unsorted input
    assert sol.merge([[2,3],[4,5],[1,2]]) == [[1,3],[4,5]], "Test 6 failed"
    
    # Test case 7: Contained intervals
    assert sol.merge([[1,10],[2,3],[4,5],[6,7]]) == [[1,10]], "Test 7 failed"
    
    # Test case 8: Multiple merges
    assert sol.merge([[1,3],[2,4],[3,5],[4,6]]) == [[1,6]], "Test 8 failed"
    
    print("✅ All tests passed!")


def trace_example():
    """Trace through merge process step by step."""
    intervals = [[1,3],[2,6],[8,10],[15,18]]
    
    print(f"\nTracing: {intervals}")
    print("="*70)
    
    # Sort first
    intervals.sort()
    print(f"After sorting: {intervals}")
    print("="*70)
    print(f"{'Step':<6} {'Current':<15} {'Next':<15} {'Action':<30}")
    print("="*70)
    
    merged = []
    current_start, current_end = intervals[0]
    
    print(f"{'0':<6} {str([current_start, current_end]):<15} {'':<15} {'Start with first interval':<30}")
    
    for i, (start, end) in enumerate(intervals[1:], 1):
        action = ""
        if start <= current_end:
            new_end = max(current_end, end)
            action = f"Overlap! Merge to [{current_start},{new_end}]"
            current_end = new_end
        else:
            action = f"No overlap. Save [{current_start},{current_end}]"
            merged.append([current_start, current_end])
            current_start, current_end = start, end
        
        print(f"{i:<6} {str([current_start, current_end]):<15} {str([start, end]):<15} {action:<30}")
    
    merged.append([current_start, current_end])
    
    print("="*70)
    print(f"Final result: {merged}")


def visualize_intervals():
    """Visualize intervals on a timeline."""
    def draw_timeline(intervals, title):
        print(f"\n{title}")
        print("="*60)
        
        # Find range
        all_nums = [num for interval in intervals for num in interval]
        min_val, max_val = min(all_nums), max(all_nums)
        
        # Draw each interval
        for i, (start, end) in enumerate(intervals):
            line = [' '] * (max_val + 1)
            for j in range(start, end + 1):
                line[j] = '█'
            print(f"Interval {i+1}: {start:2d} {' '.join(line[min_val:max_val+1])} {end:2d}")
        
        # Draw scale
        scale = [str(i % 10) for i in range(min_val, max_val + 1)]
        print(f"Scale:      {' '.join(scale)}")
    
    # Before merge
    intervals_before = [[1,3],[2,6],[8,10],[15,18]]
    draw_timeline(intervals_before, "Before Merge")
    
    # After merge
    sol = Solution()
    intervals_after = sol.merge(intervals_before)
    draw_timeline(intervals_after, "After Merge")


def edge_cases():
    """Test various edge cases."""
    sol = Solution()
    
    print("\n" + "="*50)
    print("Edge Cases:")
    print("="*50)
    
    test_cases = [
        ([[0,0]], "Single point interval"),
        ([[1,2],[2,3],[3,4]], "Chain of touching intervals"),
        ([[1,100],[2,3],[4,5]], "Small intervals inside large"),
        ([[1,4],[0,4]], "Unsorted with complete overlap"),
        ([[1,4],[0,0]], "Point and range"),
    ]
    
    for intervals, description in test_cases:
        result = sol.merge(intervals)
        print(f"{description:40} {intervals} → {result}")


if __name__ == "__main__":
    test_solution()
    trace_example()
    visualize_intervals()
    edge_cases()
